File: src-tauri/src/game/math.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::game::point::Point;
use crate::game::size::Size;

use super::consts::{PI, PI12, PI32};

#[derive(Serialize, Deserialize, Debug, Clone, TS)]
#[ts(export)]
pub struct MathFnLinear2D {
    points: Vec<Point>,
    k: Vec<f32>,
    n: Vec<f32>,
}

impl MathFnLinear2D {
    pub fn new(points: Vec<Point>) -> MathFnLinear2D {
        let mut k: Vec<f32> = vec![];
        let mut n: Vec<f32> = vec![];
        for i in 1..points.len() {
            k.push((points[i].y - points[i - 1].y) / (points[i].x - points[i - 1].x));
            n.push(points[i - 1].y - k[i - 1] * points[i - 1].x);
        }
        MathFnLinear2D { k, n, points }
    }

    pub fn calc(&self, x: f32) -> f32 {
        let mut best_section_idx = 0;
        let len = self.points.len() - 1;
        if x >= self.points[len].x {
            best_section_idx = len - 1;
        } else {
            for i in 0..len {
                if x >= self.points[i].x && x <= self.points[i + 1].y {
                    best_section_idx = i;
                    break;
                }
            }
        }
        self.k[best_section_idx] * x + self.n[best_section_idx]
    }

    pub fn inverse(&self, y: f32) -> f32 {
        let mut best_section_idx = 0;
        let len = self.points.len() - 1;
        if y >= self.points[len].y {
            best_section_idx = len - 1;
        } else {
            for i in 0..len {
                if y >= self.points[i].y && y < self.points[i + 1].y {
                    best_section_idx = i;
                    break;
                }
            }
        }
        (y - self.n[best_section_idx]) / self.k[best_section_idx]
    }
}


pub struct Math {}

impl Math {
    pub fn distance_between_points(start: &Point, end: &Point) -> f32 {
        let x = (end.x - start.x).abs();
        let y = (end.y - start.y).abs();
        (x * x + y * y).sqrt()
    }

    pub fn are_points_near(point1: &Point, point2: &Point, delta: &Size) -> bool {
        point1.x > point2.x - delta.width
            && point1.x < point2.x + delta.width
            && point1.y > point2.y - delta.height
            && point1.y < point2.y + delta.height
    }

    pub fn get_angle(position: &Point, target: &Point) -> f32 {
        let dx = target.x - position.x;
        let dz = target.y - position.y;
        let mut angle: f32 = 0.0;
        if dx == 0.0 {
            angle = PI12;
            if dz < 0.0 {
                angle = PI32;
            }
        } else if dz == 0.0 {
            if dx < 0.0 {
                angle = PI;
            }
        } else {
            angle = (dz / dx).atan();
            if dx < 0.0 && dz > 0.0 {
                angle = PI + angle;
            } else if dx < 0.0 && dz < 0.0 {
                angle = PI + angle;
            } else if dx > 0.0 && dz < 0.0 {
                angle = 2.0 * PI + angle
            }
        }
        angle
    }

    pub fn rad_to_deg(rad: f32) -> f32 {
        180.0 * rad / PI
    }

    pub fn deg_to_rad(deg: f32) -> f32 {
        PI * deg / 180.0
    }

    pub fn fn_linear_2d(points: Vec<Point>) -> MathFnLinear2D {
        MathFnLinear2D::new(points)
    }
}

File: game/math.go
package game

import (
	math2 "math"
)

type MathFnLinear2D struct {
	Points []Point   `json:"points"`
	K      []float64 `json:"k"`
	N      []float64 `json:"n"`
}

func NewMathFnLinear2D(points []Point) MathFnLinear2D {
	var k []float64
	var n []float64
	for i := 1; i < len(points); i++ {
		k = append(k, (points[i].Y-points[i-1].Y)/(points[i].X-points[i-1].X))
		n = append(n, points[i-1].Y-k[i-1]*points[i-1].X)
	}
	return MathFnLinear2D{
		Points: points,
		K:      k,
		N:      n,
	}
}

func (math *MathFnLinear2D) Calc(x float64) float64 {
	bestSectionIdx := 0
	l := len(math.Points) - 1
	if x >= math.Points[l].X {
		bestSectionIdx = l
	} else {
		for i := 0; i < l; i++ {
			if x >= math.Points[i].X && x < math.Points[i+1].X {
				bestSectionIdx = i
				break
			}
		}
	}
	return math.K[bestSectionIdx]*x + math.N[bestSectionIdx]
}

func (math *MathFnLinear2D) Inverse(y float64) float64 {
	bestSectionIdx := 0
	l := len(math.Points) - 1
	if y >= math.Points[l].Y {
		bestSectionIdx = l
	} else {
		for i := 0; i < l; i++ {
			if y >= math.Points[i].Y && y < math.Points[i+1].Y {
				bestSectionIdx = i
				break
			}
		}
	}
	return (y - math.N[bestSectionIdx]) / math.K[bestSectionIdx]
}

func MathDistanceBetweenPoint(start Point, end Point) float64 {
	x := math2.Abs(end.X - start.X)
	y := math2.Abs(end.Y - start.Y)
	return math2.Sqrt(x*x + y*y)
}

func MathArePointsNear(point1 Point, point2 Point, delta Point) bool {
	return point1.X > point2.X-delta.X && point1.X < point2.X+delta.X && point1.Y > point2.Y-delta.Y && point1.Y < point2.Y+delta.Y
}

func MathGetAngle(position Point, target Point) float64 {
	dx := target.X - position.X
	dz := target.Y - position.Y
	var angle = 0.0
	if dx == 0.0 {
		angle = PI12
		if dz < 0.0 {
			angle = PI32
		}
	} else if dz == 0.0 {
		if dx < 0.0 {
			angle = PI
		}
	} else {
		angle = math2.Atan(dz / dx)
		if dx < 0.0 && dz > 0.0 {
			angle = PI + angle
		} else if dx < 0.0 && dz < 0.0 {
			angle = PI + angle
		} else if dx > 0.0 && dz < 0.0 {
			angle = 2.0*PI + angle
		}
	}
	return angle
}

func MathRadToDeg(rad float64) float64 {
	return 180.0 * rad / PI
}

func MathDegToRad(deg float64) float64 {
	return PI * deg / 180.0
}
